embedded-graphics/src/primitives/ellipse.rs

```diff
@@ -0,0 +1,518 @@
+
+    // inner_threshold: u32,
+    inner_size: Size,
+    inner_color: Option<C>,
+
+    center: Point,
+}
+
+impl<C> StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    fn new(styled: &Styled<Ellipse, PrimitiveStyle<C>>) -> Self {
+        let has_fill_color = styled.style.fill_color.is_some();
+        let has_stroke_color = styled.style.stroke_color.is_some();
+
+        // Always use a stroke width of 0 if no stroke color was set.
+        let stroke_width = if has_stroke_color {
+            styled.style.stroke_width
+        } else {
+            0
+        };
+
+        let outer_size = styled.primitive.size;
+        let inner_size = Size::new(
+            outer_size.width.saturating_sub(2 * stroke_width),
+            outer_size.height.saturating_sub(2 * stroke_width),
+        );
+        let iter = if has_stroke_color || has_fill_color {
+            Points::new(&styled.primitive)
+        } else {
+            Points::empty()
+        };
+
+        Self {
+            iter,
+            outer_size,
+            outer_color: styled.style.stroke_color,
+            inner_size,
+            inner_color: styled.style.fill_color,
+            center: styled.primitive.center_2x(),
+        }
+    }
+}
+
+impl<C> Iterator for StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    type Item = Pixel<C>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self {
+            inner_size,
+            inner_color,
+            outer_color,
+            center,
+            ..
+        } = *self;
+
+        self.iter.find_map(|point| {
+            let inner_threshold = size_to_threshold(&inner_size, &(point * 2 - center));
+
+            let color = if inner_threshold <= 0 {
+                inner_color
+            } else {
+                outer_color
+            };
+
+            color.map(|color| Pixel(point, color))
+        })
+    }
+}
+
+impl<'a, C: 'a> Drawable<C> for &Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_iter(self)
+    }
+}
+
+/// Uses the ellipse equation b^2 * x^2 + a^2 * y^2 - a^2 * b^2 to return a value signifying whether
+/// a given point lies inside (negative), on (0) or outside (positive) an ellipse centered around
+/// `(0, 0)` with width and height defined by the `size` parameter.
+fn size_to_threshold(size: &Size, point: &Point) -> i32 {
+    let Size {
+        width: a,
+        height: b,
+    } = size;
+
+    let Point { x, y } = point;
+
+    let x = x.pow(2);
+    let y = y.pow(2);
+
+    // Special case for circles, where width and height are equal
+    if a == b {
+        let diameter = *a;
+
+        let threshold = if diameter <= 4 {
+            diameter.pow(2) - diameter
+        } else {
+            diameter.pow(2)
+        } as i32;
+
+        (x + y) - threshold
+    } else {
+        let a = a.pow(2) as i32;
+        let b = b.pow(2) as i32;
```

Comment:
These don't change and could be stored in the iter instead of `size` to improve performance.

Reply:
This would require storing and passing in the non-squared version of `a` as that's used in the optimised circle path. I tried adding that in but it's not very clean.

Reply:
Ok, I've pushed a change that adds a `diameter` argument to the function. A slight performance improvement. Let me know what you think.